### PR TITLE
Add excavation

### PR DESCRIPTION
**Pull request by mdallongeville**
- Implement the Excavation action
- Introduced the Choice concept
- Implemented a bit more the Tiles

### CODE REVIEW

File: domain/actionSpace/utils.ts [hunk 1]
```diff
@@ -33,16 +33,23 @@ function assertPlayerCanHaveMoreDwarf(player: Player) {
...
     return [
         {
             original: player,
-            diff: {
-                resources: player.resources.remove(furnishing.price),
-                tilesToPlace: [furnishing as Tile].concat(player.tilesToPlace.concat()),
-            },
+            diff: { resources: player.resources.remove(furnishing.price) },
         },
+        ...placeTile(furnishing as Tile, player),
```
Comment (yangmenghan) [inline domain/actionSpace/utils.ts:52]: 👍

File: domain/actionSpace/excavation.ts [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
+    Choice,
     EntityMutation,
     Game,
     PlayerId,
     Resources,
     ResourceType,
 } from "../entity";
-import { bookActionSpace, takeResources } from "./utils";
+import { bookActionSpace, placeTile, takeResources } from "./utils";
+import { assertIsTile } from "../entity/Tile";
```
Comment (yangmenghan) [inline domain/actionSpace/excavation.ts:12]: I wonder if we should place this into the index, too 🤔

**Review by yangmenghan — APPROVED** [review]
LGTM, Good job :D ! 💪

File: domain/entity/Tile.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+export class Tile {
+    constructor(id: TileId) {
+        this.id = id;
+    }
+    id: TileId;
+}
+
+export type TileId = CavernId | FurnishingId;
+
+export function assertIsTile(tile: unknown): asserts tile is Tile {
```
[the remark this answers is not in the document]
Comment (mdallongeville) [reply domain/entity/Tile.ts:12]: Yes, it's a nice feature :D => see https://2ality.com/2020/06/type-guards-assertion-functions-typescript.html#asserting-the-type-of-an-argument%3A-asserts-%C2%ABarg%C2%BB-is-%C2%ABtype%C2%BB for more info

**Pull request merged**